### ISSUE TITLE
[V.1.x] Product settings are not updating line items of already added items to "uncompleted" carts

### ISSUE THREAD

**Issue by alessioacella** [medusajs/medusa#7874]
### Describe the bug

When updating the product settings, such as changing a product from "discountable" to "not discountable", the line items in already existing "uncompleted" carts are not updated with the new settings. As a result, if a discount was previously applied to a product, it remains applied even after the product is set to "not discountable". This behavior is not consistent with standard ecommerce practices, where changes in product settings should reflect in all uncompleted carts.

### System information

Medusa version (including plugins): 1.20.x
Node.js version: 18/20
Database: postgresql
Operating system: Ubuntu, MacOS, Linux Ami
Browser (if relevant): 

### Steps to reproduce the behavior

1. Set a product as "discountable".
2. Add the product to a cart and apply a discount.
3. Change the product setting to "not discountable".
4. Retrieve the cart and observe that the previous discount is still applied.

### Expected behavior

Uncompleted carts should reflect the updated product settings. If a product is set to "not discountable", the line items in uncompleted carts should also update to remove any previously applied discounts.

**Comment by voxsar** [comment]
I am not sure if this is a bug? If i add something to my cart within a discount sale period and retrieve it later, i should be able to use that discount right, because i have added it to my cart?

**Comment by alessioacella** [comment]
I understand your concern, but from my experience, it's standard practice in the e-commerce industry for uncompleted carts to reflect updated product settings. This means that if a product's status is changed to "not discountable," any previously applied discounts in uncompleted carts should be removed accordingly.

I've tested various platforms, including Shopify and Drupal, and all of them follow a similar policy: uncompleted carts are dynamically updated to ensure they reflect the current product settings and discount policies. This approach ensures consistency and fairness across the shopping experience.

Allowing discounts to remain active in uncompleted carts even after the product settings have been updated can lead to discrepancies and potential financial and inventory management issues. Therefore, updating uncompleted carts to match the current product settings is crucial for maintaining a balanced and fair shopping experience for all users.

For example, imagine adding a discount code to your cart and then waiting two years to complete the order with that coupon. It sounds strange and impractical, as it undermines the purpose of time-limited promotions and discounts. This scenario highlights why it's essential for carts to reflect updated product settings and discount policies promptly.

If the intention is to allow discounts to remain valid beyond the sale period for items already in the cart, this should be explicitly stated and managed through the platform's administrative settings. However, this is an uncommon approach and could introduce challenges in maintaining overall fairness and consistency.
